Refine compile_model API

TobyRoseman opened the pull request:
* For efficiency restrict allowable model types 
* Add instructions for previously allowable types
* Add optional parameter for where to save compiled model

CI: https://gitlab.com/coremltools1/coremltools/-/pipelines/1075605326

YifanShenSZ commented on coremltools/models/utils.py:
```diff
@@ -1035,25 +1034,31 @@ def compile_model(model: _Union['_ct.models.MLModel', str, _Model_pb2.Model]) ->
 
-    # Check parameter
+    # Check model parameter
     if not isinstance(model, (str, _Model_pb2.Model, _ct.models.MLModel)):
```
I understand the `if not isinstance(model, (str, _Model_pb2.Model, _ct.models.MLModel))` is kind of for "backward compatibility", but imho it would confuse new comers. How about
```
if isinstance(model, str):
    raise Exception
if isinstance(model, _ct.models.MLModel):
    raise Exception
if not isinstance(model, _Model_pb2.Model):
    raise Exception
```

YifanShenSZ commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -88,3 +93,18 @@ def test_non_default_compute_units(self):
+            dst_path = temp_dir + "/foo.badFileExtension"
+            with pytest.raises(Exception) as e:
+                compiled_model_path = compile_model(self.spec, dst_path)
+            assert " file extension." in str(e.value)
```
Nice unit test coverage 💯

aseemw commented on coremltools/models/utils.py:
```diff
@@ -1004,24 +1004,23 @@ def _convert_to_float(feature):
-        MLModel : Instantiated Core ML model to compile
+    destination_path: str
+        Path where the compiled model will be saved.
 
```
Maybe good to add a sample code showing the usage of the API, like done for some of the other APIs

TobyRoseman replied:
Done.

jakesabathia2 commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -55,8 +57,10 @@ def test_file_input(self):
-            compiled_model_path = compile_model(spec_file_path)
-            self._test_compile_model_path(compiled_model_path)
+
+            with pytest.raises(Exception) as e:
```
will recommend to use `with pytest.raises(Exception, match=...)`

jakesabathia2 commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -66,8 +70,9 @@ def test_spec_input(self):
         ml_model = MLModel(self.spec)
-        compiled_model_path = compile_model(ml_model)
-        self._test_compile_model_path(compiled_model_path)
+        with pytest.raises(Exception) as e:
```
same: will recommend to use with pytest.raises(Exception, match=...)

TobyRoseman commented:
Updated CI: https://gitlab.com/coremltools1/coremltools/-/pipelines/1088430866

TobyRoseman commented:
CI after rebase: https://gitlab.com/coremltools1/coremltools/-/pipelines/1088549487

junpeiz commented on coremltools/models/utils.py:
```diff
@@ -1035,25 +1062,31 @@ def compile_model(model: _Union['_ct.models.MLModel', str, _Model_pb2.Model]) ->
-        raise Exception("Compiling a Core ML models is only support on macOS 10.13 or higher.")
+    # Check model parameter
+    if isinstance(model, str):
+        raise Exception("To get a compiled model from a saved MLModel, first load the model, "
```
nit: Try to use more specific exception to avoid the general Exception as suggested by PEP8. I would recommend `ValueError` for invalid parameters in util functions.

junpeiz commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -88,3 +91,27 @@ def test_non_default_compute_units(self):
+        # Check bad input
+        with TemporaryDirectory() as temp_dir:
+            dst_path = temp_dir + "/foo.badFileExtension"
+            with pytest.raises(Exception) as e:
```
nit: Change the general `Exception` to the corresponding specific exception (ValueError) as suggested in another comment in `utils.py`.

TobyRoseman replied:
Done.

TobyRoseman commented:
@aseemw - this API only works for very simple mlprograms, i.e. mlprograms without a weight/blob file. Having this API take a weight file would be a relatively easy change. However that doesn't really solve the problem. Currently, we don't give user any functionality to build their own weight.

I agree having this API support `mlprogram` would be great. However I think that's a much bigger conversation, than this PR. This PR was intended to just be a small refinement of the existing API.

TobyRoseman commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -51,12 +53,13 @@ def _test_compile_model_path(self, compiled_model_path, compute_units=ComputeUni
+            file_path = save_dir + '/m.mlmodel'
+            MLModel(self.spec).save(file_path)
+
+            with pytest.raises(Exception, match=", first load the model, ") as e:
```
Sure. Done.

TobyRoseman commented on coremltools/test/neural_network/test_compiled_model.py:
```diff
@@ -88,3 +91,27 @@ def test_non_default_compute_units(self):
+            dst_path = temp_dir + "/foo.badFileExtension"
+            with pytest.raises(Exception) as e:
+                compiled_model_path = compile_model(self.spec, dst_path)
+            assert " file extension." in str(e.value)
```
Done

TobyRoseman commented:
I've addressed feedback from @junpeiz. I have also added a doc string comment saying `compile_model` does not support mlprograms which use a blob file.

Updated CI: https://gitlab.com/coremltools1/coremltools/-/pipelines/1099285623
